bmc/oem/dell.go

```diff
@@ -60,3 +65,370 @@ func (r *Dell) GetTaskMonitorDetails(ctx context.Context, taskMonitorResponse *h
+						k,
+						BMCDellAttributes,
+					)
+			}
+		}
+	}
+
+	filteredAttr, err := d.getFilteredBMCRegistryAttributes(false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(filteredAttr) == 0 {
+		return nil, fmt.Errorf("'ManagerAttributeRegistry' not found")
+	}
+
+	// from the gives attributes to change, find the ones which can be changed and get current value for them
+	result := make(redfish.SettingsAttributes, len(attributes))
+	var errs []error
+	for _, name := range attributes {
+		if entry, ok := filteredAttr[name]; ok {
+			// enumerations current setting comtains display name.
+			// need to be checked with the actual value rather than the display value
+			// as the settings provided will have actual values.
+			// replace display values with actual values
+			if strings.ToLower(string(entry.Type)) == string(redfish.EnumerationAttributeType) {
+				for _, attrValue := range entry.Value {
+					if attrValue.ValueDisplayName == mergedBMCAttributes[name] {
+						result[name] = attrValue.ValueName
+						break
+					}
+				}
+				if _, ok := result[name]; !ok {
+					errs = append(
+						errs,
+						fmt.Errorf(
+							"current setting '%v' for key '%v' not found in possible values for it (%v)",
+							mergedBMCAttributes[name],
+							name,
+							entry.Value,
+						))
+				}
+			} else {
+				result[name] = mergedBMCAttributes[name]
+			}
+		} else {
+			// possible error in settings key
+			errs = append(errs, fmt.Errorf("setting key '%v' not found in possible settings", name))
+		}
+	}
+	if len(errs) > 0 {
+		return result, fmt.Errorf(
+			"some errors found in the settings '%v'.\nPossible settings %v",
+			errs,
+			maps.Keys(filteredAttr),
+		)
+	}
+
+	return result, nil
+}
+
+func (d *DellIdracManager) UpdateBMCAttributesApplyAt(
+	attrs redfish.SettingsAttributes,
+	applyTime common.ApplyTime,
+) error {
+
+	BMCattributeValues, err := d.getCurrentBMCSettingAttribute()
+	if err != nil {
+		return err
+	}
+
+	payloads := make(map[string]redfish.SettingsAttributes, len(BMCattributeValues))
+	for key, value := range attrs {
+		for _, eachAttr := range BMCattributeValues {
+			if _, ok := eachAttr.Attributes[key]; ok {
+				if data, ok := payloads[eachAttr.Settings.SettingsObject.String()]; ok {
+					data[key] = value
+				} else {
+					payloads[eachAttr.Settings.SettingsObject.String()] = make(redfish.SettingsAttributes)
+					payloads[eachAttr.Settings.SettingsObject.String()][key] = value
+				}
+				// keys cant be duplicate. Hence, break once its already found in one of idrac settings sub types
+				break
+			}
+		}
+	}
+
+	// If there are any allowed updates, try to send updates to the system and
+	// return the result.
+	if len(payloads) > 0 {
+		var errs []error
+		// for wach sub type, apply the settings
+		for settingPath, payload := range payloads {
+			// fetch the etag required for settingPath
+			etag, err := func(uri string) ([]string, error) {
```

Comment:
Is it possible to batch the requests instead of doing a loop? If yes, it would be a good follow-up.